src/core/unit.c

```diff
@@ -3195,6 +3243,14 @@ int unit_add_dependency_by_name(Unit *u, UnitDependency d, const char *name, boo
         if (u->manager && FLAGS_SET(u->manager->test_run_flags, MANAGER_TEST_RUN_IGNORE_DEPENDENCIES))
                 return 0;
 
+        r = unit_is_recursive_template_dependency(u, name);
+        if (r < 0)
+                return r;
+        if (r == 1) {
```

Comment:
i think it would be slightly nicer to check for `> 0` rather than `== 1` here, in the usual sense we have in place for tri-state bools (which this is not, but conceptually close)